### PR TITLE
fix app panics during start for AddRESTHandlers

### PR DESCRIPTION
**Pull request by srinivas365**
Closes: https://github.com/gofr-dev/gofr/issues/858

### CODE REVIEW

File: pkg/gofr/crud_handlers.go [hunk 1]
```diff
@@ -61,7 +61,16 @@ type entity struct {
 
 // scanEntity extracts entity information for CRUD operations.
 func scanEntity(object interface{}) (*entity, error) {
-	entityType := reflect.TypeOf(object).Elem()
+	if object == nil {
+		return nil, errInvalidObject
```
Comment (Umang01-hash) [inline pkg/gofr/crud_handlers.go:65]: Can we find some way to include the struct details ,
So thinking of a scenario where user has used method `AddRESTHandlers` for 2 different structs, one he has passed by value and other by pointer.

So when we will run the code, w err message : ` invalid object for AddRESTHandlers` will not give clear description of which object is invalid....

**Review by Umang01-hash — CHANGES_REQUESTED** [review]

File: pkg/gofr/crud_handlers.go [hunk 2]
```diff
@@ -61,7 +63,16 @@ type entity struct {
...
 // scanEntity extracts entity information for CRUD operations.
 func scanEntity(object interface{}) (*entity, error) {
-	entityType := reflect.TypeOf(object).Elem()
+	if object == nil {
+		return nil, errObjectIsNil
+	}
+
+	objType := reflect.TypeOf(object)
+	if objType.Kind() != reflect.Ptr {
+		return nil, errObjectNotPointer
```
Comment (Umang01-hash) [inline pkg/gofr/crud_handlers.go:72]: Can't we use `objType.Name()` here for a specialized error message in **errObjectNotPointer**

A error message like `failed to register routes for user struct, as it is passed by value should be pointer`

will be more informative from user's point of view

**Review by Umang01-hash — CHANGES_REQUESTED** [review]

File: pkg/gofr/crud_handlers.go (hunk 2 above)
Comment (srinivas365) [reply pkg/gofr/crud_handlers.go]: Updated the error message and included the struct name as well

File: pkg/gofr/crud_handlers.go [hunk 3]
```diff
@@ -10,8 +10,10 @@ import (
 )
 
 var (
-	errInvalidObject  = errors.New("unexpected object given for AddRESTHandlers")
-	errEntityNotFound = errors.New("entity not found")
+	errInvalidObject    = errors.New("unexpected object given for AddRESTHandlers")
+	errEntityNotFound   = errors.New("entity not found")
+	errObjectNotPointer = errors.New("pointer object is not given for AddRESTHandlers")
```
Comment (srijan-27) [inline pkg/gofr/crud_handlers.go:15]: this is unused now, since we have changed the error for not pointer case.

File: pkg/gofr/crud_handlers.go [hunk 4]
```diff
@@ -61,7 +63,16 @@ type entity struct {
...
 // scanEntity extracts entity information for CRUD operations.
 func scanEntity(object interface{}) (*entity, error) {
-	entityType := reflect.TypeOf(object).Elem()
+	if object == nil {
+		return nil, errObjectIsNil
+	}
+
+	objType := reflect.TypeOf(object)
+	if objType.Kind() != reflect.Ptr {
+		return nil, fmt.Errorf("failed to register routes for %s struct, as it is passed by value should be pointer", objType.Name())
```
Comment (srijan-27) [inline pkg/gofr/crud_handlers.go:72]: we can just mention that passed object is not pointer, no need to specify `passed by value`
also, we can present struct in quotes like: `'%s' struct`

**Review by srijan-27 — CHANGES_REQUESTED** [review]

File: pkg/gofr/crud_handlers.go (hunk 4 above)
Comment (srijan-27) [reply pkg/gofr/crud_handlers.go]: @srinivas365 - The linter error is yet to be fixed.

File: pkg/gofr/crud_handlers.go [hunk 5]
```diff
@@ -61,7 +63,16 @@ type entity struct {
...
 // scanEntity extracts entity information for CRUD operations.
 func scanEntity(object interface{}) (*entity, error) {
-	entityType := reflect.TypeOf(object).Elem()
+	if object == nil {
+		return nil, errObjectIsNil
+	}
+
+	objType := reflect.TypeOf(object)
+	if objType.Kind() != reflect.Ptr {
+		return nil, fmt.Errorf("failed to register routes for '%s' struct, %w", objType.Name(), errNonPointerObject)
```
Comment (puellanivis) [inline pkg/gofr/crud_handlers.go:72]: Use the formatting verb `%q` not `'%s'`.

**Review by vipul-rawat — APPROVED** [review]
`

**Review by Umang01-hash — APPROVED** [review]

**Review by srijan-27 — APPROVED** [review]

**Pull request merged**